### PR TITLE
Emily dev

### PR DESCRIPTION
**Pull request by emily-tsai11**
- bin/nano_files_2016.py: updated signal sample directories + 1 whitespace change
- bin/submitJobs.py: changed run time from 10 to 60 minutes + whitespace/capitalisation changes
- helper/Count_Ntotal.py: syntax and whitespace changes
- helper/bkg_estimate: added functionality to make a background estimate using a 2D histogram and the ABCD method
- helper/cutflow.py: removed constants from the script and updated it to pull constants from drawHists.py + capitalisation changes
- hists/hadd.py: whitespace changes
- include/MyAnalysis.h: added 2D histograms, added a new char_to_int method for converting WPs + whitespace changes
- include/event_candidate.h: whitespace changes
- include/lepton_candidate.h: added WP and decayMode lepton attributes + 1 whitespace change
- plot/drawHists.py: reformatting constants, added 2D histograms, added background estimate
- plot/plotter.py: whitespace changes, added 2D histogram plotting method, added method for plotting background estimate comparisons
- src/MyAnalysis.cc: adding and filling 2D histograms, added indexing for the regions + whitespace changes
- src/event_candidate.cc: whitespace changes
- src/jet_candidate.cc: 1 whitespace change
- src/lepton_candidate.cc: added WP and decayMode lepton attributes
- src/main.cc: adding DY events for testing + 1 whitespace change

### CODE REVIEW

File: helper/cutflow.py
Comment (Jingyan95) [inline helper/cutflow.py:1]: I'd suggest to only include backgrounds when the percentage numbers are calculated. Also, would it be possible to add more significant figures when yields are small but nonzero?

File: bin/submitJobs.py [hunk 1]
```diff
@@ -53,7 +53,7 @@
 if data_2018:
     SAMPLES.update(nano_files_2018.data2018_samples)
 
-jobruntime = 600 # 10min
+jobruntime = 3600 # 60 min
```
Comment (Jingyan95) [inline bin/submitJobs.py:56]: does this mean the features from fake tau estimate will increase the run-time? or this is a more "playing safe".

File: bin/nano_files_2016.py [hunk 2]
```diff
@@ -11,13 +11,13 @@
 ##cut += " && ((Sum$(Tau_pt>18 && abs(Tau_eta)<2.3 && Tau_idDeepTau2017v2p1VSe>=1 && Tau_idDeepTau2017v2p1VSmu>=1 && Tau_idDeepTau2017v2p1VSjet>=1 && Tau_decayMode!=5 && Tau_decayMode!=6))>=1)"
 
 ##Scalar interaction
-mc2016_samples['2016_LFVStScalarU'] = [['/eos/cms/store/user/jingyan/LFV_Trilep_Inclusive_May1_BtagSF/2016_LFVStScalarU_UL/CRAB_UserFiles/crab_Trilep_Inclusive_May1_BtagSF_2016_LFVStScalarU_UL/230501_150328/0000/'], 'mc','LFVStScalarU','2016', '','0.417' ,'16.81','100000',100]
-mc2016_samples['2016_LFVTtScalarU'] = [['/eos/cms/store/user/jingyan/LFV_Trilep_Inclusive_May1_BtagSF/2016_LFVTtScalarU_UL/CRAB_UserFiles/crab_Trilep_Inclusive_May1_BtagSF_2016_LFVTtScalarU_UL/230501_145549/0000/'], 'mc','LFVTtScalarU','2016', '','0.012','16.81','100000',100]
+mc2016_samples['2016_fr_LFVStScalarU'] = [['/eos/user/e/etsai/workspace/MCProduction/SMEFTfr_ST_clequ1_emutu/root_final'], 'mc','LFVStScalarU', '2016', '', '0.417', '16.81', '115000', 115]
```
Comment (Jingyan95) [inline bin/nano_files_2016.py:14]: I think we do need a TT sample for completeness

File: plot/plotter.py [hunk 3]
```diff
@@ -718,10 +785,138 @@ def SummaryPlot(hists, SignalHists, Fnames, reg = "region", regName = ["region",
...
+        os.makedirs(folder + '/' + year + '/' + c + '/' + ch)
+    if not os.path.exists(folder + '/' + year + '/' + c + '/' + ch + '/' + reg):
+        os.makedirs(folder + '/' + year + '/' + c + '/' + ch + '/' + reg)
+    canvas.Print(folder + '/' + year + '/' + c + '/' + ch + '/' + reg + '/' + var + '_' + fname[0:-5] + ".pdf")
+
+    del canvas
+    gc.collect()
+
+
+def getLumi(year):
```
Comment (Jingyan95) [inline plot/plotter.py:912]: Very neat trick!

File: helper/cutflow.py [hunk 4]
```diff
@@ -1,79 +1,44 @@
...
+        for numreg, namereg in enumerate(regions):
+            for c in range(len(charges)):
+                for ch in range(len(channels)):
+                    h = Hists[numyear][numf][c][ch][numreg][15].Clone()
                     for b in range(nBins):
-                        Counts[f][reg][b][0] += h.GetBinContent(b + 1)
-                        Counts[f][reg][b][1] += h.GetBinError(b + 1)
-                        if 'Data' not in f: Totals[reg][b] += h.GetBinContent(b + 1)
-
+                        Counts[namef][namereg][b][0] += h.GetBinContent(b + 1)
```
Comment (Jingyan95) [inline helper/cutflow.py:30]: due to low stats, sometimes NLO samples can give negative event counts. so it would be good to check & send negative values to zero

File: helper/cutflow.py [hunk 5]
```diff
@@ -120,7 +85,7 @@
     outFile.write('    \\end{enumerate}\n')
     outFile.write('  \\end{frame}\n')
 
-    # cutflow table slides, 1 per region
+    # Cutflow table slides, 1 per region
```
Comment (Jingyan95) [inline helper/cutflow.py:88]: Rather than using "makebox", I would recommend something like "\resizebox{0.98\linewidth}{!}{%". This makes automatically adjust size of the table and make sure it does not go outside of the frame

File: helper/cutflow.py [hunk 6]
```diff
@@ -1,79 +1,44 @@
...
-# set up an argument parser
-parser = argparse.ArgumentParser()
-parser.add_argument('--n', dest = 'NAMETAG', default = '2016')
-ARGS = parser.parse_args()
-name = ARGS.NAMETAG
-
-sys.path.append(os.path.join(os.path.dirname(sys.path[0]), 'bin'))
-loc = os.path.dirname(sys.path[0]) + '/'
-HistAddress = loc + 'hists/'
+nBins = 18 # Number of bins in subSR histogram
```
Comment (Jingyan95) [inline helper/cutflow.py:3]: this nBins number can be obtained through GetNbinX()
